### PR TITLE
feat: adding hardhat

### LINKED ISSUE #625: Post transactions from the simulator to the hardhat network
Post the rollup transactions to the hardhat network instead of to a database table

### PR DESCRIPTION
**Pull request by kstroobants**
Fixes #625 

# What

- Add Hardhat node service with Docker configuration
- Add necessary dependencies for contract compilation setup
- Add access to compiled hardhat contract in jsonrpc service
- Configure hardhat.config.js with local network settings and hardhat specifics
- Add 1 test account with 10,000 ETH and have free transactions so no funding is needed
- Add web3 Python package to backend requirements file
- Update .gitignore for Hardhat artifacts and cache
- Add solidity contract to write transactions to that emits events
- Added database migration file so that a transaction has the ghost contract address of the hardhat network
- When a genlayer contract is deployed, then a hardhat contract is deployed, both are linked in the CurrentState table
- When a genlayer write method is executed, then the new transaction gets the hardhat contract from the CurrentState table
- When a genlayer transaction status changes, then a rollup transaction is created on the hardhat network
- Add hardhat port in .env
- Remove rollup transactions database table

# Why

- To add more value to the user
- See the rollup transactions on a test Ethereum network

# Testing done

- Tested by running the simulator
- Add hardhat test with code from the transaction_processor

# Decisions made

At the moment one hardhat account is used.

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

- Run the simulator
- Run the unit test

# User facing release notes

The simulator contains the hardhat network development environment

### CODE REVIEW

File: backend/database_handler/transactions_processor.py [hunk 1]
```diff
@@ -128,6 +138,64 @@ def insert_transaction(
...
+                "hardhat/artifacts/contracts/GhostContract.sol/GhostContract.json",
+            )
+
+            with open(contract_file, "r") as f:
+                contract_json = json.loads(f.read())
+                abi = contract_json["abi"]
+                bytecode = contract_json["bytecode"]
+
+            # Create the contract instance
+            contact = self.web3.eth.contract(abi=abi, bytecode=bytecode)
```
Comment (cristiam86) [inline backend/database_handler/transactions_processor.py:159]:
```suggestion
            contract = self.web3.eth.contract(abi=abi, bytecode=bytecode)
```

**Review by cristiam86 — COMMENTED** [review]
Good job. Let's try to see if we can have the same addresses  for both contracts (genlayer contract and ghost contract)

File: examples/contracts/amm_adaptive.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
+import json
```
Comment (cristiam86) [inline examples/contracts/amm_adaptive.py:1]: @mpaya5 nice to add this one, but let's do it on its own PR please and also with the new GenVM syntax

**Review by cristiam86 — APPROVED** [review]
Minor comment about amm intelligent contract example

File: docker/Dockerfile.hardhat [hunk 3]
```diff
@@ -0,0 +1,20 @@
+FROM node:20.11-alpine3.19
+
+WORKDIR /app
+
+RUN apk add --no-cache python3 make g++ netcat-openbsd
```
[the remark this answers is not in the document]
Comment (mpaya5) [reply docker/Dockerfile.hardhat:5]: I will fix that in the next commit

File: examples/contracts/amm_adaptive.py (hunk 2 above)
Comment (mpaya5) [reply examples/contracts/amm_adaptive.py:1]: I deleted the amm_adaptive.py

File: backend/database_handler/transactions_processor.py (hunk 1 above)
Comment (mpaya5) [reply backend/database_handler/transactions_processor.py]: Done

**Comment by mpaya5** [comment]
image
image

Are we certain that deleting the `RollupTransaction` is the correct approach?

Since we made these changes, I've noticed that once the transaction is accepted, the methods to interact with the contract no longer appears in the frontend, even after reloading the page.

Could this be related to the removal of `RollupTransactions`? Let me know your thoughts @kstroobants @cristiam86
